File: 2816-double-a-number-represented-as-a-linked-list/2816-double-a-number-represented-as-a-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def solve(self, head):
        if head == None:
            return 0
        carry = self.solve(head.next)
        curr_val = head.val * 2 + carry
        head.val = curr_val % 10
        return curr_val // 10

    def doubleIt(self, head):
        carry = self.solve(head)
        if carry != 0:
            new_head = ListNode(carry)
            new_head.next = head
            return new_head
        
        return head
        

File: 2816-double-a-number-represented-as-a-linked-list/2816-double-a-number-represented-as-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private int solve(ListNode head) {
        if(head == null) return 0;
        int carry = solve(head.next);

        int currVal = head.val * 2 + carry;
        head.val = currVal % 10;
        return currVal / 10;
    }
    
    public ListNode doubleIt(ListNode head) {
        int carry = solve(head);
        if(carry != 0) {
            ListNode newHead = new ListNode(carry);
            newHead.next = head;
            return newHead;
        }

        return head;
    }
}